components/raftstore/src/store/peer.rs

```diff
@@ -2786,6 +2800,55 @@ where
         fail_point!("after_send_to_apply_1003", self.peer_id() == 1003, |_| {});
     }
 
+    /// Check long uncommitted proposals and log some info to help find why.
+    pub fn check_long_uncommitted_proposals<T>(&mut self, ctx: &mut PollContext<EK, ER, T>) {
+        if self.has_long_uncommitted_proposals(ctx) {
+            self.log_for_long_uncommitted_proposals()
+        }
+    }
+
+    /// Try to log enough info to help find why there exists long uncommitted proposals.
+    fn log_for_long_uncommitted_proposals(&mut self) {
+        let status = self.raft_group.status();
+        let mut buffer: Vec<(u64, u64, u64)> = Vec::new();
+        if let Some(prs) = status.progress {
+            for (id, p) in prs.iter() {
+                buffer.push((*id, p.commit_group_id, p.matched));
+            }
+        }
+        warn!(
+            "found long uncommitted proposals";
+            "region_id" => self.region_id,
+            "peer_id" => self.peer.get_id(),
+            "progress" => ?buffer,
+            "cache_first_index" => ?self.get_store().get_entry_cache_first_index(),
+        );
+    }
+
+    /// Check if there is long uncommitted proposal.
+    fn has_long_uncommitted_proposals<T>(&mut self, ctx: &mut PollContext<EK, ER, T>) -> bool {
+        let mut has_long_uncommitted = false;
```

Comment:
if commit index is close to latest index, early return to avoid getting time too often.

Reply:
Changed. Check if it has uncommitted log first, and then do this check.